File: prueba-diagnostica-lenguajes/Problema 1/problema1.py
"""
Problema 1 - Triángulo de Pascal y Evaluación de Polinomio (x+1)^n
Curso: Lenguajes y Compiladores - UNEG
Implementación: Python con memoria dinámica (listas dinámicas)
"""

import time
import sys


def format_number(num):
    """
    Formatea un número para mostrarlo como entero si es entero, de lo contrario como float.
    """
    if num == int(num):
        return str(int(num))
    else:
        return str(num)


def generar_coeficientes(n):
    """
    Genera los coeficientes del polinomio (x+1)^n usando el Triángulo de Pascal.
    Usa listas de Python como equivalente a memoria dinámica.
    Retorna una lista con los n+1 coeficientes.
    """
    # Caso base
    if n == 0:
        return [1]

    # Inicializamos con la primera fila del triángulo: [1]
    fila_anterior = [1]

    for i in range(1, n + 1):
        # Creamos la nueva fila dinámicamente
        fila_actual = [0] * (i + 1)
        fila_actual[0] = 1          # Primer elemento siempre es 1
        fila_actual[i] = 1          # Último elemento siempre es 1

        # Los elementos intermedios se calculan sumando los dos de arriba
        for j in range(1, i):
            fila_actual[j] = fila_anterior[j - 1] + fila_anterior[j]

        fila_anterior = fila_actual  # La fila actual pasa a ser la anterior

    return fila_anterior


def mostrar_polinomio(coeficientes, n):
    """
    Construye y muestra la representación textual del polinomio.
    Ejemplo para n=3: x^3 + 3x^2 + 3x + 1
    """
    terminos = []

    for i, coef in enumerate(coeficientes):
        # El exponente del término actual va de n hasta 0
        exponente = n - i

        if coef == 0:
            continue  # Saltamos términos con coeficiente 0

        if exponente == 0:
            # Término independiente (sin x)
            terminos.append(str(coef))
        elif exponente == 1:
            # Término lineal
            if coef == 1:
                terminos.append("x")
            else:
                terminos.append(f"{coef}x")
        else:
            # Término con potencia
            if coef == 1:
                terminos.append(f"x^{exponente}")
            else:
                terminos.append(f"{coef}x^{exponente}")

    return " + ".join(terminos) if terminos else "0"


def evaluar_polinomio_pasos(coeficientes, n, x):
    """
    Evalúa f(x) = (x+1)^n usando los coeficientes generados.
    Muestra el proceso paso a paso:
    Usa el Esquema de Horner para evaluar eficientemente.
    f(x) = c0*x^n + c1*x^(n-1) + ... + cn
    """
    print(f"\n{'='*55}")
    print(f"  Evaluación paso a paso de f({format_number(x)}) = ({format_number(x)}+1)^{n}")
    print(f"{'='*55}")

    if n == 0:
        print(f"  f({x}) = 1")
        return 1

    print(f"\n  Polinomio: f(x) = {mostrar_polinomio(coeficientes, n)}")
    print(f"\n  Sustituyendo x = {format_number(x)}:")
    print()

    resultado = 0
    terminos_evaluados = []

    for i, coef in enumerate(coeficientes):
        exponente = n - i
        valor_termino = coef * (x ** exponente)
        terminos_evaluados.append(valor_termino)

        if exponente == 0:
            print(f"    Término {i+1}: {coef} × {format_number(x)}^0 = {coef} × 1 = {format_number(valor_termino)}")
        elif exponente == 1:
            print(f"    Término {i+1}: {coef} × {format_number(x)}^1 = {coef} × {format_number(x)} = {format_number(valor_termino)}")
        else:
            print(f"    Término {i+1}: {coef} × {format_number(x)}^{exponente} = {coef} × {format_number(x**exponente)} = {format_number(valor_termino)}")

        resultado += valor_termino

    print()
    suma_str = " + ".join(format_number(t) for t in terminos_evaluados)
    print(f"  Suma total: {suma_str}")
    print(f"\n  ∴ f({format_number(x)}) = {format_number(resultado)}")
    print(f"  Verificación directa: ({format_number(x)}+1)^{n} = {format_number((x+1)**n)}")
    print(f"{'='*55}")

    return resultado


def medir_tiempo_n100():
    """
    Mide el tiempo de ejecución para n=100 y guarda resultados en archivo.
    """
    n = 100
    print(f"\nMidiendo tiempo de ejecución para n={n}...")

    inicio = time.perf_counter()
    coeficientes = generar_coeficientes(n)
    fin = time.perf_counter()

    tiempo_ms = (fin - inicio) * 1000

    # Guardamos en archivo TXT
    with open("resultado_n100.txt", "w", encoding="utf-8") as archivo:
        archivo.write(f"Resultado para n={n}\n")
        archivo.write(f"{'='*60}\n")
        archivo.write(f"Tiempo de ejecución: {tiempo_ms:.6f} ms\n")
        archivo.write(f"Número de coeficientes generados: {len(coeficientes)}\n\n")
        archivo.write("Coeficientes (desde x^100 hasta x^0):\n")
        for i, c in enumerate(coeficientes):
            archivo.write(f"  C({i}) = {c}\n")

    print(f"  Tiempo: {tiempo_ms:.6f} ms")
    print(f"  Resultados guardados en 'resultado_n100.txt'")

    return tiempo_ms, coeficientes


def main():
    print("=" * 55)
    print("  PROBLEMA 1: Triángulo de Pascal - (x+1)^n")
    print("  Lenguajes y Compiladores - UNEG")
    print("=" * 55)

    # --- Parte a): Generar y mostrar el polinomio ---
    try:
        n = int(input("\nIngrese el valor de n (entero no negativo): "))
        if n < 0:
            print("Error: n debe ser no negativo.")
            sys.exit(1)
    except ValueError:
        print("Error: ingrese un número entero válido.")
        sys.exit(1)

    print(f"\n  Generando coeficientes para (x+1)^{n}...\n")
    coeficientes = generar_coeficientes(n)

    print(f"  Coeficientes obtenidos (Triángulo de Pascal, fila {n}):")
    print(f"  {coeficientes}")

    polinomio = mostrar_polinomio(coeficientes, n)
    print(f"\n  Polinomio resultante:")
    print(f"  f(x) = (x+1)^{n} = {polinomio}")

    # --- Parte b): Evaluar el polinomio para x dado ---
    try:
        x = float(input(f"\nIngrese el valor de x para evaluar f(x) = (x+1)^{n}: "))
    except ValueError:
        print("Error: ingrese un número válido.")
        sys.exit(1)

    evaluar_polinomio_pasos(coeficientes, n, x)

    # --- Medición de tiempo para n=100 ---
    print("\n" + "=" * 55)
    print("  MEDICIÓN DE TIEMPO PARA n=100")
    print("=" * 55)
    medir_tiempo_n100()

    print("\n¡Programa finalizado exitosamente!")


if __name__ == "__main__":
    main()

File: prueba-diagnostica-lenguajes/Problema 1/problema1.c
/*
 * Problema 1 - Triangulo de Pascal y Evaluacion de Polinomio (x+1)^n
 * Curso: Lenguajes y Compiladores - UNEG
 * Implementacion: C con memoria dinamica real (malloc / free)
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

/* ===========================================================
 *  FUNCION: generar_coeficientes
 *  Recibe : n (grado del polinomio)
 *  Retorna: puntero a arreglo dinamico con n+1 coeficientes
 *           El llamador es responsable de liberar la memoria
 * ===========================================================
 *
 *  MEMORIA DINAMICA:
 *  En cada fila del triangulo pedimos al sistema operativo
 *  exactamente la memoria que necesitamos con malloc().
 *  La fila anterior se libera con free() una vez que ya
 *  no la necesitamos, evitando fugas de memoria.
 *
 *  Representacion interna:
 *
 *  fila_anterior -> [ 1 ]                    (n=0)
 *  fila_actual   -> [ 1 | 1 ]                (n=1)
 *  fila_actual   -> [ 1 | 2 | 1 ]            (n=2)
 *  fila_actual   -> [ 1 | 3 | 3 | 1 ]        (n=3)
 *                      ^               ^
 *                   malloc()        free() al terminar
 */
long long* generar_coeficientes(int n) {

    /* Caso base: (x+1)^0 = 1 */
    if (n == 0) {
        long long* base = (long long*) malloc(sizeof(long long));
        if (!base) { fprintf(stderr, "Error: malloc fallo\n"); exit(1); }
        base[0] = 1;
        return base;
    }

    /* Empezamos con la fila 0 del triangulo: [1] */
    long long* fila_anterior = (long long*) malloc(sizeof(long long));
    if (!fila_anterior) { fprintf(stderr, "Error: malloc fallo\n"); exit(1); }
    fila_anterior[0] = 1;

    long long* fila_actual = NULL;

    for (int i = 1; i <= n; i++) {

        /* Pedimos memoria para i+1 elementos (la fila i tiene i+1 numeros) */
        fila_actual = (long long*) malloc((i + 1) * sizeof(long long));
        if (!fila_actual) { fprintf(stderr, "Error: malloc fallo\n"); exit(1); }

        /* Primer y ultimo elemento de cada fila siempre son 1 */
        fila_actual[0] = 1;
        fila_actual[i] = 1;

        /* Los elementos del medio: suma de los dos de arriba */
        for (int j = 1; j < i; j++) {
            fila_actual[j] = fila_anterior[j - 1] + fila_anterior[j];
        }

        /* Ya no necesitamos la fila anterior: liberamos su memoria */
        free(fila_anterior);

        /* La fila actual se convierte en la anterior para la proxima iteracion */
        fila_anterior = fila_actual;
    }

    return fila_actual;  /* Devolvemos el puntero, el llamador hara free() */
}


/* ===========================================================
 *  FUNCION: mostrar_polinomio
 *  Imprime en pantalla la expresion algebraica del polinomio
 *  Ejemplo: [1,3,3,1] con n=3 -> "x^3 + 3x^2 + 3x + 1"
 * =========================================================== */
void mostrar_polinomio(long long* coef, int n) {
    int primero = 1;  /* Bandera para no imprimir "+" antes del primer termino */

    for (int i = 0; i <= n; i++) {
        int exponente = n - i;
        long long c = coef[i];

        if (c == 0) continue;

        /* Separador entre terminos */
        if (!primero) printf(" + ");
        primero = 0;

        if (exponente == 0) {
            /* Termino independiente: solo el numero */
            printf("%lld", c);
        } else if (exponente == 1) {
            /* Termino lineal */
            if (c == 1) printf("x");
            else        printf("%lldx", c);
        } else {
            /* Termino con potencia */
            if (c == 1) printf("x^%d", exponente);
            else        printf("%lldx^%d", c, exponente);
        }
    }
    printf("\n");
}


/* ===========================================================
 *  FUNCION: potencia_entera
 *  Calcula base^exp de forma iterativa (sin usar pow())
 *  para trabajar con enteros exactos
 * =========================================================== */
double potencia(double base, int exp) {
    double resultado = 1.0;
    for (int i = 0; i < exp; i++) resultado *= base;
    return resultado;
}

int es_entero(double valor) {
    long long entero = (long long) valor;
    return fabs(valor - entero) < 1e-9;
}

void imprimir_numero(double valor) {
    if (es_entero(valor))
        printf("%lld", (long long) valor);
    else
        printf("%.4f", valor);
}


/* ===========================================================
 *  FUNCION: evaluar_polinomio_pasos
 *  Evalua f(x) termino a termino mostrando cada paso
 * =========================================================== */
double evaluar_polinomio_pasos(long long* coef, int n, double x) {
    printf("\n=======================================================\n");
    printf("  Evaluacion paso a paso de f(");
    imprimir_numero(x);
    printf(") = (x+1)^%d\n", n);
    printf("=======================================================\n");

    if (n == 0) {
        printf("  f(");
        imprimir_numero(x);
        printf(") = 1\n");
        return 1.0;
    }

    printf("\n  Sustituyendo x = ");
    imprimir_numero(x);
    printf(":\n\n");

    double resultado = 0.0;

    for (int i = 0; i <= n; i++) {
        int exp      = n - i;
        double xpow  = potencia(x, exp);
        double valor = coef[i] * xpow;
        resultado   += valor;

        if (exp == 0) {
            printf("    Termino %2d: %lld x^0 = %lld x 1 = ", i+1, coef[i], coef[i]);
            imprimir_numero(valor);
            printf("\n");
        } else if (exp == 1) {
            printf("    Termino %2d: %lld x^1 = %lld x ", i+1, coef[i], coef[i]);
            imprimir_numero(x);
            printf(" = ");
            imprimir_numero(valor);
            printf("\n");
        } else {
            printf("    Termino %2d: %lld x^%d = %lld x ", i+1, coef[i], exp, coef[i]);
            imprimir_numero(xpow);
            printf(" = ");
            imprimir_numero(valor);
            printf("\n");
        }
    }

    printf("\n  Resultado f(");
    imprimir_numero(x);
    printf(") = ");
    imprimir_numero(resultado);
    printf("\n");
    printf("  Verificacion directa: (");
    imprimir_numero(x);
    printf("+1)^%d = ", n);
    imprimir_numero(potencia(x + 1, n));
    printf("\n");
    printf("=======================================================\n");

    return resultado;
}


/* ===========================================================
 *  FUNCION: medir_y_guardar_n100
 *  Mide el tiempo para n=100 y escribe los resultados en
 *  un archivo .txt junto con el tiempo de Python para comparar
 * =========================================================== */
void medir_y_guardar_n100(double tiempo_python_ms) {
    int n = 100;

    /* --- Medicion de tiempo con clock() --- */
    clock_t inicio = clock();

    long long* coef = generar_coeficientes(n);

    clock_t fin = clock();

    double tiempo_ms = (double)(fin - inicio) / CLOCKS_PER_SEC * 1000.0;

    printf("\n=======================================================\n");
    printf("  MEDICION DE TIEMPO PARA n=%d\n", n);
    printf("=======================================================\n");
    printf("  Tiempo C      : %.6f ms\n", tiempo_ms);
    printf("  Tiempo Python : %.6f ms\n", tiempo_python_ms);
    if (tiempo_python_ms > 0)
        printf("  C es %.1fx mas rapido que Python\n",
               tiempo_python_ms / tiempo_ms);
    printf("  Coeficientes generados: %d\n", n + 1);

    /* --- Escritura en archivo TXT --- */
    FILE* archivo = fopen("resultado_n100.txt", "a");  /* append: agrega al existente */
    if (!archivo) {
        fprintf(stderr, "Error: no se pudo abrir resultado_n100.txt\n");
        free(coef);
        return;
    }

    fprintf(archivo, "\n");
    fprintf(archivo, "========================================\n");
    fprintf(archivo, "RESULTADO EN C para n=%d\n", n);
    fprintf(archivo, "========================================\n");
    fprintf(archivo, "Tiempo de ejecucion C      : %.6f ms\n", tiempo_ms);
    fprintf(archivo, "Tiempo de ejecucion Python : %.6f ms\n", tiempo_python_ms);
    if (tiempo_python_ms > 0)
        fprintf(archivo, "C es %.1fx mas rapido que Python\n",
                tiempo_python_ms / tiempo_ms);
    fprintf(archivo, "\nCoeficientes de (x+1)^%d:\n", n);
    for (int i = 0; i <= n; i++) {
        fprintf(archivo, "  C(%3d) = %lld\n", i, coef[i]);
    }
    fclose(archivo);

    printf("  Resultados agregados en 'resultado_n100.txt'\n");
    printf("=======================================================\n");

    free(coef);  /* IMPORTANTE: liberamos la memoria pedida con malloc() */
}


/* ===========================================================
 *  FUNCION PRINCIPAL
 * =========================================================== */
int main() {
    printf("=======================================================\n");
    printf("  PROBLEMA 1: Triangulo de Pascal - (x+1)^n\n");
    printf("  Lenguajes y Compiladores - UNEG  [version C]\n");
    printf("=======================================================\n");

    /* --- Parte a): Pedir n y generar coeficientes --- */
    int n;
    printf("\nIngrese el valor de n (entero no negativo): ");
    if (scanf("%d", &n) != 1 || n < 0) {
        fprintf(stderr, "Error: n debe ser un entero no negativo.\n");
        return 1;
    }

    printf("\n  Generando coeficientes para (x+1)^%d...\n\n", n);

    /* Llamada con memoria dinamica */
    long long* coeficientes = generar_coeficientes(n);

    /* Mostrar la fila del triangulo */
    printf("  Coeficientes (fila %d del Triangulo de Pascal):\n  [ ", n);
    for (int i = 0; i <= n; i++) {
        printf("%lld", coeficientes[i]);
        if (i < n) printf(", ");
    }
    printf(" ]\n");

    /* Mostrar el polinomio */
    printf("\n  Polinomio resultante:\n");
    printf("  f(x) = (x+1)^%d = ", n);
    mostrar_polinomio(coeficientes, n);

    /* --- Parte b): Evaluar para x dado --- */
    double x;
    printf("\nIngrese el valor de x para evaluar f(x) = (x+1)^%d: ", n);
    if (scanf("%lf", &x) != 1) {
        fprintf(stderr, "Error: ingrese un numero valido.\n");
        free(coeficientes);
        return 1;
    }

    evaluar_polinomio_pasos(coeficientes, n, x);

    /* Liberamos la memoria de los coeficientes del polinomio */
    free(coeficientes);

    /* --- Medicion n=100 y comparacion con Python --- */
    double tiempo_python = 0.261558;  /* Tiempo medido previamente en Python */
    medir_y_guardar_n100(tiempo_python);

    printf("\n¡Programa finalizado exitosamente!\n\n");
    return 0;
}
